### PR TITLE
Add spectrum support

### PR DESCRIPTION
**Pull request by WTFox**
Hello! :wave: 🙂 We are picking up where #243 left off.

This pull request's main goal was to pass the tests without changing any of @mtami's and @briandailey's excellent work. My changes consisted of only formatting and removing python 3.5 in tests. Initially, I wanted to recreate the previous failure, but since python 3.5 is end-of-life, I couldn't even get it installed on my 2019 MBP running Ventura. I created https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/issues/262 to decide which python versions we wish to maintain support for.

Tests are passing locally.
image

Please let me know if I've missed anything!

## Todos
- [x] MIT compatible
- [x] Tests
- [x] Documentation
- [x] Updated CHANGES.rst

### CODE REVIEW

**Comment by WTFox** [comment]
I updated CHANGES.rst to reflect that we no longer support Python 3.5 and SQLAlchemy 1.3. If that's okay, I need to probably update setup.py, too. If so, could you provide the values you'd like for the following attributes:

- `install_requires`
- `python_requires`
- `classifiers`

**Comment by jasonamyers** [comment]
are these deprecations enough that there should be a major version bump?

**Comment by Brooke-white** [comment]
@jasonamyers, thats a good question. I think a major version bump would be needed for dropping sqlalchemy 1.3 support..though I am unsure how other libraries have handled dropping support for unsupported versions of python. Any idea?

For some comparison to [other sqlalchemy dialects](https://docs.sqlalchemy.org/en/20/dialects/), I browsed a few and saw many still support sqlalchemy 1.3. 

@WTFox , Do we know if this PR explicitly breaks sqlalchemy 1.3 support? I'd be interested to see how this PR tests for sqlalchemy 1.3. Per [sqlalchemy's 1.3.23 release](https://github.com/sqlalchemy/sqlalchemy/blob/rel_1_3_23/setup.py#L178) it should support python 3.4-3.9, so let's try running against a newer not-unsupported Python version and see what results we get.

If we do see breaking tests, unsure how much effort we want to place into supporting a legacy version of sqlalchemy if the fixes are non-trivial. On a related note, sqlalchemy 2.0 was recently released, and I will open an issue for supporting that in the future. But it is for another day :)

tagging @graingert here to get his opinion on this.

**Comment by WTFox** [comment]
@Brooke-white Good call! I reinstated tests to test against SQLAlchemy version 1.3 using python 3.9. I also updated the PR body and CHANGES.rst to reflect this.

**Comment by jasonamyers** [comment]
@WTFox how would you feel about adding tox tests for 3.10 for the higher rev sqlalchemy versions?

**Comment by WTFox** [comment]
@jasonamyers I added tests for the following:

- Python 3.10 - SQLA 1.4
- Python 3.10 - SQLA 2.0

This actually surfaced some changes in pytest parametrized tests. I updated the tests in test_reflection.py to utilize the newer `pytest.param` function. See [this StackOverflow answer for context(https://stackoverflow.com/a/30575822).

**Comment by WTFox** [comment]
Quick update:

After adding SQLAlchemy 2.0 tests, a few issues surfaced that I'm working through now.

File: sqlalchemy_redshift/dialect.py
Comment (Brooke-white) [inline sqlalchemy_redshift/dialect.py:666]: do we want to print or log here? idk if we have existing log statements in the project, but it could be good to introduce one in this case

**Review by Brooke-white — COMMENTED** [review]
this may be out of scope for this PR...but figured i'd mention as you're working in these methods incase the effort to do this is minimal:

when executing queries, could we use bind variables  in-place of string concatenation? we'd want to have some logic to set the paramstyle to some common type (e.g. format), then re-set to the original value post query execution. only caveat with this is we cant use bind variables for sub-queries (i dont think i saw any but there may be some), due to limitations on Redshift server.

I know we're not using bind parameters in the existing code base, but we _really_ should be  as it provides an added level of protection against sql injection. if it's too much extra effort for now just let me know and I can open an issue for addressing this in the future :)

**Comment by WTFox** [comment]
This PR is ready for review. We decided against adding SQLAlchemy 2.0 compatibility. There is a private SQLAlchemy dialect function that we're using, `_get_column_info` that no longer exists in the dialect, and the effort required to fix that exceeds the scope of this work. See [this](https://docs.sqlalchemy.org/en/20/changelog/whatsnew_20.html#major-architectural-performance-and-api-enhancements-for-database-reflection).

Tests are passing for the following tox envs:
image

A few noteworthy items:
- docs were failing, but pinning sphinx to 1.6 fixed it
- support for python 3.11 was added in psycopg2 2.9, so we're using that for the python 3.11 tests
- alembic was pinned at a higher version to fix some errors as well.

**Comment by WTFox** [comment]
> if it's too much extra effort for now just let me know and I can open an issue for addressing this in the future :)

If it works for you, I'd prefer to open a new issue so that this PR can move forward. :)

File: sqlalchemy_redshift/dialect.py
Comment (WTFox) [reply sqlalchemy_redshift/dialect.py:666]: Logging would be better, I can change this.

**Comment by Brooke-white** [comment]
sounds good -- opened #265 wrt the bind parameter changes. As our CI is broken right now, I'll kick off these tests on my end and confirm everything looks good. I should get back to you before end of week with results. thanks again for taking on adding spectrum support! :)

**Comment by WTFox** [comment]
> thanks again for taking on adding spectrum support! :)

I'm just standing on the shoulders of @jasonamyers @briandailey and @mtami :)

**Comment by Brooke-white** [comment]
Hey @WTFox -- I ran the CI test suite on my local machines and encountered a couple failures...it looks like a bunch since we run with so many configurations, but with a quick look I just saw 3 tests failing.

1. `tests/test_reflection.py::test_external_table_reflection`
2. `tests/test_reflection.py::test_reflection`
3. `tests/test_long_tablename.py::test_long_tablename`

Some quick analysis on each...
1. we try to execute some sql using role `'arn:aws:iam::123456789012:role/mySpectrumRole'`. this execution fails because the role does not exist in the account where my test cluster is. but it will not exist in any account, since it's an example role :)...so we will need to create an IAM role, which isn't an issue...but I would like to avoid exposing the role ARN (as it includes the AWS account number), so if we could use some means similar to what we use for `PGPASSWORD` e.g. env var to pass this value into the test framework to dynamically pass to the test framework, I think we could accomplish this. very much open to other ideas though :)
2. The expected sql isn't generated. I didn't look at this one too much, but it appears some clauses may be missing.
3. hits some "database is being accessed by others"


for 3, I remember this error from the prior PR. you can read [the convo there](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/pull/243#issuecomment-1027109938) for some context. Prior analysis by mtami@ pointed towards transaction isolation level as a potential culprit. 

I've attached output from my run of the tests via `tox` below -- I know this is far from ideal as we should really have our CI up and running, but a means to and end until then :)

For debugging these I'd recommend to checkout my convo with mtami@ in the previous PR and to consider using a Redshift trial period for some lightweight debugging of the few tests for which issue is seen. 


[spectrum-support-tox.txt](https://github.com/sqlalchemy-redshift/sqlalchemy-redshift/files/10690154/spectrum-support-tox.txt)

**Comment by WTFox** [comment]
@Brooke-white Thanks for the follow-up! Next steps for me will be to set up a redshift instance and see if I can reproduce those integration test failures locally. That might take a bit, but I'll try to keep you updated along the way.

**Comment by WTFox** [comment]
Happy Friday!

With the help of the wonderful @jasonamyers, I got this test suite up and running against our redshift instance. This was made possible by adding a few pytest fixtures with default values but allowing environment variables to override them. Here's a list of environment variables supported:

- PGPASSWORD
- REDSHIFT_HOST
- REDSHIFT_PORT
- REDSHIFT_USERNAME
- REDSHIFT_DATABASE
- AWS_ACCOUNT_ID
- REDSHIFT_IAM_ROLE_NAME

IAM role ARNs are constructed using these values. I've written quite a bit for each new fixture in conftest.py, but putting these in the readme or a contributing doc might be a good idea. Where would be the best place for documentation on running the tests?

We also fixed a bug in the reflection query that _might_ have been the cause of the "DB is being accessed by others" issue. The entire test suite passes when I specify a single tox environment, but I got an OutOfMemory exception when running them all in parallel. Can you kick off a CI run at your convenience?

Finally, I removed the environments for python 3.11 and psycopg 2.9 as this would add extra work to the scope of this PR.

Please take a look and let me know if anything else stands out. Thank you for the help! 🙂

**Comment by Brooke-white** [comment]
Hi @WTFox -- apologies for the delay. Taking a look at this PR now.

File: tests/conftest.py
Comment (Brooke-white) [inline tests/conftest.py:82]: could we add some details around how this role should be configured? e.g. what permissions does it grant? is it associated with the cluster?

File: tests/conftest.py
Comment (Brooke-white) [inline tests/conftest.py:136]: similar comment to above -- what permissions do the roles have, when do you want to define this env var, are these roles associated with the cluster?

**Comment by Brooke-white** [comment]
> I've written quite a bit for each new fixture in conftest.py, but putting these in the readme or a contributing doc might be a good idea. Where would be the best place for documentation on running the tests?

Thank you for making all of these improvements to our test suite! I think a section in the read me would be helpful. I added some comments wrt the role(s) that could be addressed in the readme. also, if there is any pre-test setup (e.g. table creation) required on the cluster, or region restriction that would be useful details also.

> The entire test suite passes when I specify a single tox environment, but I got an OutOfMemory exception when running them all in parallel. Can you kick off a CI run at your convenience?

Working on this now...our CI is still broken, so I'm working on setting up something temporary in my aws account so tests can at least be kicked off via a remote machine. In parallel, I'm running these tests on my local machine.

**Review by Brooke-white — COMMENTED** [review]
overall look good -- i will add some more comments once I wrap up getting the test suite kicked off locally + on a remote machine

File: tests/conftest.py
Comment (Brooke-white) [inline tests/conftest.py:81]: for this logic around the role name & constructing the role arn from various pieces, could we rework this to use look for an env var e.g. `REDSHIFT_IAM_ROLE_ARN` that is the whole arn e.g. `arn:aws:iam::12345678912:role/ExampleRole`

**Comment by Brooke-white** [comment]
>  I got an OutOfMemory exception when running them all in parallel. Can you kick off a CI run at your convenience?

Was this issue seen using `tox --parallel`? At least so far, I haven't been able to repo it using a very much "built in a half-day" CI setup 😄 . Maybe the OOM is specific to some limits hit on your machine that are not hit on the ec2 instance used in the CI? I'll keep trying to repro...

**Comment by WTFox** [comment]
> Was this issue seen using `tox --parallel`?

Yeah, I was getting a successful test run when I'd specify a single env. I'm almost 100% certain this is due to my personal computer constraints.

File: tests/conftest.py
Comment (WTFox) [reply tests/conftest.py:82]: Hmm. Unfortunately, since we were testing this, we gave the role all permissions to Redshift. However, we did have to associate the role to the cluster so I'll update this docstring and the one below to reflect this.

File: tests/conftest.py
Comment (WTFox) [reply tests/conftest.py:81]: This was actually by choice. Some tests look for the ARN, and others look for the account_id and role_name. Passing in an ARN that doesn't reflect the account_id and role_name causes a few test failures due to the mismatch. 

The alternative would be to pass in the role arn and have the account_id, and role name fixtures parse that to get their values. If that's preferred, I can change my implementation.

Comment (Brooke-white) [reply tests/conftest.py:81]: Sounds good, yes please update when you get a chance :)

**Comment by Brooke-white** [comment]
@WTFox -- I'm considering moving our CI test execution to non-parallelized, due to hitting different sorts of errors when run in parallel that are specifically related to the tests running in parallel i.e. external schema already exists because it was created in another test session running at the same time. I don't want to beat these parallelized test issues into the ground, since it has more to do with the test framework than your change, and want to go ahead and get your changes merged once the doc updates + changes around ARN are made :) 

fwiw i believe running tests in parallel was a cost based decision as the previous maintainer was using a personal AWS account. given we don't run this CI too often, i think it should be alright to move to non-parallelized tests at least in the short term.

I'll do a final pass through our CI once the mentioned changes are pushed and then expect to get this merged!

**Comment by WTFox** [comment]
Sounds good, I'll work on this tomorrow. Thank you for working through this with me!

**Comment by Brooke-white** [comment]
taking a look -- will run through CI

**Comment by WTFox** [comment]
> taking a look -- will run through CI

I have a local test run that's almost done, it should be done any minute now. It might be best to hold off until I get that result. I can tag you when ready. Sorry for the lack of communication there. Updating the README is still on my todo, too. But I can do that while CI is running.

**Comment by Brooke-white** [comment]
👍 sounds good -- I'll hold off until all else is pushed

**Comment by WTFox** [comment]
@Brooke-white, my test run passed. Kick off the CI run at your convenience, and I'll start updating the README.

**Review by Brooke-white — APPROVED** [review]

**Comment by Brooke-white** [comment]
Thank you @WTFox @jasonamyers @mtami and all else who helped with this change! I am unavailable next week, but anticipate making a release before end of this month.

**Comment by matthewgdv** [comment]
Hi all, just wanted to mention a regression added in the PR that has forced me to pin `sqlalchemy-redshift==0.8.12` in my organization's library code repo. This happens specifically when using a `future=True` engine in sqlalchemy1.4, which previously worked fine. This is not the same as using sqlalchemy2.0, which has made large code changes to its connection and reflection logic relative to 1.4 (even with `FutureEngine`) that break `sqlalchemy-redshift` in other ways.

Basically, this PR added usages of `connection.connect()`  in the reflection code without checking if the connection was already open, which against a `FutureEngine` causes an error because connection forking is no longer allowed from sqlalchemy2.0 onwards.

I can open a new issue if that's better for tracking purposes?

**Comment by Brooke-white** [comment]
Hi @matthewgdv , yes please open a new issue so we can work to fix this

**Pull request merged**
